File: src/finance_analysis/interfaces/api/v1/schemas/backtests.py
"""Pydantic contracts for the unified backtest API."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator


class BacktestConfigRequest(BaseModel):
    engine: str
    strategy_key: str
    market: str
    code: str
    start_date: date
    end_date: date
    parameters: dict[str, Any] = Field(default_factory=dict)
    benchmark_code: str | None = None

    @field_validator("engine", "strategy_key")
    @classmethod
    def lower_key(cls, value: str) -> str:
        return value.strip().lower()

    @field_validator("market", "code", "benchmark_code")
    @classmethod
    def upper_key(cls, value: str | None) -> str | None:
        return value.strip().upper() if value else value


class BacktestRunCreate(BacktestConfigRequest):
    initial_cash: float = Field(default=100000, gt=0)


class EngineResponse(BaseModel):
    key: str
    name: str
    description: str
    display_order: int
    is_default: bool
    recommended: bool
    available: bool
    unavailable_reason: str | None
    version: str | None
    supported_markets: list[str] | tuple[str, ...]
    supported_strategies: list[str] | tuple[str, ...]


class StrategyParameterResponse(BaseModel):
    key: str
    name: str
    type: str
    default: int
    minimum: int
    maximum: int


class StrategyResponse(BaseModel):
    key: str
    name: str
    description: str
    version: str
    frequency: str
    supported_markets: list[str] | tuple[str, ...]
    supported_engines: list[str] | tuple[str, ...]
    parameters: list[StrategyParameterResponse] | tuple[StrategyParameterResponse, ...]


class SymbolResponse(BaseModel):
    id: int
    market: str
    code: str
    name: str
    lot_size: int | None = None


class PreflightResponse(BaseModel):
    ready: bool
    engine: str
    engine_version: str | None
    strategy_key: str
    market: str
    code: str
    available_date_from: date | None
    available_date_to: date | None
    requested_trading_days: int
    available_trading_days: int
    coverage_ratio: float
    warmup_days: int
    warnings: list[str]
    errors: list[str]


class BacktestRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uid: int
    task_id: str | None
    engine: str
    engine_version: str | None
    engine_config: dict[str, Any]
    strategy_key: str
    strategy_name: str
    strategy_version: str
    market: str
    symbol_id: int
    code: str
    start_date: date
    end_date: date
    initial_cash: float
    benchmark_code: str | None
    parameters: dict[str, Any]
    price_mode: str
    market_rule_version: str
    status: str
    progress: int
    summary: dict[str, Any]
    warnings: list[str]
    error: str | None
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None


class BacktestRunListResponse(BaseModel):
    items: list[BacktestRunResponse]
    total: int
    page: int
    page_size: int


class BacktestTradeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_id: int
    code: str
    engine_order_id: str | None
    side: str
    signal_date: date
    order_date: date
    trade_date: date
    quantity: float
    price: float
    gross_amount: float
    commission: float
    tax: float
    other_fee: float
    total_fee: float
    cash_after: float
    position_after: float
    return_pct: float | None
    pnl: float | None
    exit_reason: str | None


class BacktestTradeListResponse(BaseModel):
    items: list[BacktestTradeResponse]
    total: int
    page: int
    page_size: int


class BacktestEquityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    trading_date: date
    cash: float
    position_value: float
    total_equity: float
    benchmark_equity: float | None
    daily_return_pct: float
    cumulative_return_pct: float
    benchmark_return_pct: float | None
    drawdown_pct: float


class BacktestEquityListResponse(BaseModel):
    items: list[BacktestEquityResponse]

File: web/src/types/backtests.ts
export type BacktestEngineKey = 'backtrader' | 'rqalpha';
export type BacktestMarket = 'US' | 'CN' | 'HK';
export type BacktestStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';

export type BacktestEngine = {
  key: BacktestEngineKey;
  name: string;
  description: string;
  displayOrder: number;
  isDefault: boolean;
  recommended: boolean;
  available: boolean;
  unavailableReason: string | null;
  version: string | null;
  supportedMarkets: BacktestMarket[];
  supportedStrategies: string[];
};

export type StrategyParameter = {
  key: string;
  name: string;
  type: 'integer';
  default: number;
  minimum: number;
  maximum: number;
};

export type BacktestStrategy = {
  key: string;
  name: string;
  description: string;
  version: string;
  frequency: string;
  supportedMarkets: BacktestMarket[];
  supportedEngines: BacktestEngineKey[];
  parameters: StrategyParameter[];
};

export type BacktestSymbol = {
  id: number;
  market: BacktestMarket;
  code: string;
  name: string;
  lotSize: number | null;
};

export type BacktestConfig = {
  engine: BacktestEngineKey;
  strategyKey: string;
  market: BacktestMarket;
  code: string;
  startDate: string;
  endDate: string;
  initialCash: number;
  benchmarkCode: string | null;
  parameters: Record<string, number>;
};

export type BacktestPreflight = {
  ready: boolean;
  engine: BacktestEngineKey;
  engineVersion: string | null;
  strategyKey: string;
  market: BacktestMarket;
  code: string;
  availableDateFrom: string | null;
  availableDateTo: string | null;
  requestedTradingDays: number;
  availableTradingDays: number;
  coverageRatio: number;
  warmupDays: number;
  warnings: string[];
  errors: string[];
};

export type BacktestSummary = {
  initialCash?: number;
  finalEquity?: number;
  totalReturnPct?: number;
  annualizedReturnPct?: number;
  benchmarkReturnPct?: number;
  excessReturnPct?: number;
  maxDrawdownPct?: number;
  sharpeRatio?: number | null;
  volatilityPct?: number;
  tradeCount?: number;
  winRatePct?: number;
};

export type BacktestRun = {
  id: number;
  uid: number;
  taskId: string | null;
  engine: BacktestEngineKey;
  engineVersion: string | null;
  engineConfig: Record<string, unknown>;
  strategyKey: string;
  strategyName: string;
  strategyVersion: string;
  market: BacktestMarket;
  symbolId: number;
  code: string;
  startDate: string;
  endDate: string;
  initialCash: number;
  benchmarkCode: string | null;
  parameters: Record<string, number>;
  priceMode: string;
  marketRuleVersion: string;
  status: BacktestStatus;
  progress: number;
  summary: BacktestSummary;
  warnings: string[];
  error: string | null;
  createdAt: string;
  startedAt: string | null;
  finishedAt: string | null;
};

export type BacktestTrade = {
  id: number;
  runId: number;
  code: string;
  engineOrderId: string | null;
  side: 'buy' | 'sell';
  signalDate: string;
  orderDate: string;
  tradeDate: string;
  quantity: number;
  price: number;
  grossAmount: number;
  commission: number;
  tax: number;
  otherFee: number;
  totalFee: number;
  cashAfter: number;
  positionAfter: number;
  returnPct: number | null;
  pnl: number | null;
  exitReason: string | null;
};

export type BacktestEquity = {
  tradingDate: string;
  cash: number;
  positionValue: number;
  totalEquity: number;
  benchmarkEquity: number | null;
  dailyReturnPct: number;
  cumulativeReturnPct: number;
  benchmarkReturnPct: number | null;
  drawdownPct: number;
};

export type BacktestRunList = { items: BacktestRun[]; total: number; page: number; pageSize: number };
